### PR TITLE
New Design for Transmit Notification

### PR DESCRIPTION
**Pull request by Soldia1138**
**When merged this pull request will:**
- Change the Design for our Transmit-Notification in lower-left corner
- This PR focuses on three key points: Fresh look, more space and possibility to extend notifications (e.g. for optional notifications on incoming transmissions, especially for intercom systems)
- Notification when transmitting via radio stays as long as the transmission is live 
- Add a predefined set of colors for notifications 
- Add the color coding to the documentation

### CODE REVIEW

**Comment by Drofseh** [comment]
If you're taking suggestions, I'd love to see the `displayName` or `shortName` of the vehicle rack that is being transmitted on.

**Comment by NouberNou** [comment]
Pictures?

**Comment by Soldia1138** [comment]
@Drofseh this will most probably be in another PR. However I want to design the system in a way that not only transmit notifications are available.

We agreed internally on a color coding plus Text/Icons to distinguish between different information types

**Comment by Soldia1138** [comment]
Almost forgot to post the video. Here is a first implementation of stackable notifications within the new system. Color is configurable as well. There is still a slightly annoying behaviour where only the lowest notification is overridden on an overflow. But I assume that situations with more than 5 notifications shouldn't occur very often

https://www.youtube.com/watch?v=VzI3R173G6g

**Comment by xetra11** [comment]
beautiful !

File: addons/sys_list/fnc_displayHint.sqf [hunk 1]
```diff
@@ -19,10 +23,26 @@
 GVAR(hintTitle) = _this select 0;
 GVAR(hintLine1) = _this select 1;
 GVAR(hintLine2) = _this select 2;
-GVAR(hintDuration) = 1;
-if ((count _this) == 4) then {
-    GVAR(hintDuration) = _this select 3;
+private _hintDuration = -1;
```
Comment (TheMagnetar) [inline addons/sys_list/fnc_displayHint.sqf]: Should we change these checks to params and default values
`params ["_hintTitle", "_hintLine1", "_hintLine2", ["_hintDuration", -1], ["_hintColor",[1, 0.8, 0, 1]]];`
?

File: addons/sys_list/fnc__showHintBox.sqf [hunk 2]
```diff
@@ -1,23 +1,66 @@
...
-params ["_hintIDD"];
+params ["_hintIDD", "_type", "_bufferpointer"];
 
-(_hintIDD displayCtrl 19100) ctrlSetText GVAR(hintTitle);
-(_hintIDD displayCtrl 19101) ctrlSetText GVAR(hintLine1);
-(_hintIDD displayCtrl 19102) ctrlSetText GVAR(hintLine2);
+if (_type == 0) then {
+    private _color = GVAR(hintColor);
+    _color set [3, 1];
+    (_hintIDD displayCtrl 19105) ctrlSetBackgroundColor _color;
```
Comment (TheMagnetar) [inline addons/sys_list/fnc__showHintBox.sqf]: Substituting 19105 with a define would be better to maintain the code later on. Same for all other displayCtrl Ids

File: addons/sys_core/initSettings.sqf [hunk 3]
```diff
@@ -127,3 +127,24 @@
     true,
     {[_this] call EFUNC(api,setRevealToAI)}
 ] call CBA_Settings_fnc_init;
+
+// Notification Settings
```
Comment (TheMagnetar) [inline addons/sys_core/initSettings.sqf:131]: Should we pursue this method, or should we use the notification system on the upper left corner and change the active color to green when intercom/racks are active?

File: addons/sys_list/fnc_displayHint.sqf (hunk 1 above)
Comment (Soldia1138) [reply addons/sys_list/fnc_displayHint.sqf]: Is it more efficient to to params first and assign it to GVARs later on than to assign it to GVARs directly?
```
params ["_hintTitle", "_hintLine1", "_hintLine2", ["_hintDuration", -1], ["_hintColor",[1, 0.8, 0, 1]]];

GVAR(hintTitle) = _hintTitle;
GVAR(hintLine1) = _hintLine1;
GVAR(hintLine2) = _hintLine2;
GVAR(hintColor) = _hintColor;
```

File: addons/sys_list/fnc__showHintBox.sqf (hunk 2 above)
Comment (Soldia1138) [reply addons/sys_list/fnc__showHintBox.sqf]: Agree

File: addons/sys_list/fnc_displayHint.sqf (hunk 1 above)
Comment (TheMagnetar) [reply addons/sys_list/fnc_displayHint.sqf]: From slack discussion, we use params.

File: addons/sys_core/fnc_speaking.sqf [hunk 4]
```diff
@@ -147,13 +147,14 @@ if !(GVAR(keyedMicRadios) isEqualTo []) then {
...
                 // acre_player sideChat format["_volumeModifier: %1 %2", _volumeModifier];
-                {
+               {
                     _on = [_x select 1, "getOnOffState"] call EFUNC(sys_data,dataEvent);
                     if (_on == 1) then {
                         _x params ["_unit","","_signalData","_params"];
                         if (!HASH_HASKEY(_compiledParams, netId _unit)) then {
                             HASH_SET(_compiledParams, netId _unit, []);
                         };
+                        //[format ["RX: %1", _unit], "", "", 2, [ACRE_NOTIFICATION_YELLOW]] call EFUNC(sys_list,displayHint);
```
Comment (TheMagnetar) [inline addons/sys_core/fnc_speaking.sqf]: Remove comment.

**Review by TheMagnetar — CHANGES_REQUESTED** [review]

**Review by TheMagnetar — APPROVED** [review]

File: addons/sys_list/fnc_hideHint.sqf [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * Hides a given notification layer
+ *
+ * Arguments:
+ * 0: Notification layer <STRING>
+ *
+ * Return Value:
+ * None
+ *
+ * Example:
+ * ["acre_sys_list_hintLayer_0"] call acre_sys_list_fnc_hideHint;
```
Comment (jonpas) [inline addons/sys_list/fnc_hideHint.sqf]: No semi-colon.

File: addons/sys_list/fnc_displayHint.sqf [hunk 6]
```diff
@@ -1,28 +1,48 @@
...
+ * 3: Optional display duration <NUMBER>(default: 1)
+ * 4: Optional color in RGBA format <ARRAY>(default: [1, 0.8, 0, 1])
  *
  * Return Value:
- * RETURN VALUE <TYPE>
+ * Name of cutRsc layer <STRING>
  *
  * Example:
- * [ARGUMENTS] call acre_COMPONENT_fnc_FUNCTIONNAME
+ * ["Title Line", "Line 1", "Line 2", 1, [1, 1, 1, 1]] call acre_sys_list_fnc_displayHint;
```
Comment (jonpas) [inline addons/sys_list/fnc_displayHint.sqf]: No semi-colon.

File: addons/sys_list/script_dialog_defines.hpp [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#define IDD_DISPLAYBACKGROUND 19098
+#define IDD_DISPLAY 19099
+
+#define IDC_CONTROLBACKGROUNDCOLOR 19105
+#define IDC_CONTROLBACKGROUNDBLACK 19106
+
+#define IDC_CONTROLTITLE 19100
+#define IDC_CONTROLLINE1 19101
+#define IDC_CONTROLLINE2 19102
+
```
Comment (jonpas) [inline addons/sys_list/script_dialog_defines.hpp]: Empty line too many.

**Review by jonpas — CHANGES_REQUESTED** [review]
Why not make notifications UI movable UI elements like Rack UI is?

**Comment by Soldia1138** [comment]
Need to check one more thing before finishing this

**Review by jonpas — APPROVED** [review]

**Comment by jonpas** [comment]
I would really like to make it movable.

File: addons/sys_list/CfgUIGrids.hpp [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+                        "(((safezoneW / safezoneH) min 1.2) / 1.2) / 25"
+                    };
+                };
+            };
+        };
+
+        class Variables {
+            class grid_ACRE_notification {
+                displayName = CSTRING(NotificationGrid);
+                description = CSTRING(NotificationGridDesc);
```
Comment (jonpas) [inline addons/sys_list/CfgUIGrids.hpp:23]: Those 2 strings are missing.

**Review by jonpas — CHANGES_REQUESTED** [review]
Missing strings, otherwise looks good!

File: addons/sys_list/stringtable.xml [hunk 9]
```diff
@@ -31,5 +31,13 @@
             <Chinesesimp>打开无线电</Chinesesimp>
             <Portuguese>Abrir Rádio</Portuguese>
         </Key>
+        <Key ID="STR_ACRE_sys_list_NotificationGrid">
+            <English>ACRE2 - Radio Notifications</English>
+            <German>ACRE2 - Funkbenachrichtigungen</German>
+        </Key>
+        <Key ID="STR_ACRE_sys_list_NotificationGridDesc">
+            <English>Shows notifications regarding radios and transmissions</English>
```
Comment (jonpas) [inline addons/sys_list/stringtable.xml]: Missing final dot.

**Review by jonpas — CHANGES_REQUESTED** [review]

File: addons/sys_list/CfgUIGrids.hpp (hunk 8 above)
Comment (Soldia1138) [reply addons/sys_list/CfgUIGrids.hpp:23]: Done https://github.com/IDI-Systems/acre2/pull/576/commits/920dc7e3c9dd4fa6a04bf97177295cc83521b115

**Review by jonpas — APPROVED** [review]

**Pull request merged**